File: code/Android_App/ITS-Mobile-App/app/src/main/java/pt/isel/meic/tfm/itsmobileapp/security_protocols/dlapp/tamper_proof_device/hash_chain/HashChainModuleImpl.kt
package pt.isel.meic.tfm.itsmobileapp.security_protocols.dlapp.tamper_proof_device.hash_chain

import android.util.Log
import pt.isel.meic.tfm.itsmobileapp.security_protocols.dlapp.TAG_DLAPP
import pt.isel.meic.tfm.itsmobileapp.security_protocols.dlapp.models.HashChain
import java.security.MessageDigest

/**
 * Prototype implementation of the Hash Chain Module Interface
 */
class HashChainModuleImpl(
    private val hashAlgorithm: String
) : IHashChainModule {

    override fun createHashChain(seed: ByteArray, size: Int): HashChain {

        val md = MessageDigest.getInstance(hashAlgorithm)
        val hashChain = Array<ByteArray>(size) { byteArrayOf() }

        var toDigest : ByteArray = seed

        for (i in 1..size) {
            toDigest = md.digest(toDigest)
            hashChain[i - 1] = toDigest

            //Log.v(TAG_DLAPP, "Hash idx $i: ${toDigest.contentToString()}")

            /*//Log.d(TAG_DLAPP, "Hash idx $i (hex): " + toDigest.joinToString(separator = "") {
                String.format("%02X", it)
            })*/
        }

        return HashChain(hashChain)
    }

}

File: code/MessageGenerator/security_protocols/dlapp/tamper_proof_device/hash_chain_module_impl.py
import hashlib
import logging

from security_protocols.dlapp.models.hash_chain import HashChain
from common.util import create_signed_byte_array_from


class HashChainModuleImpl:
    """
    Prototype implementation of the Hash Chain Module Interface
    """

    def __init__(self, hash_algorithm):
        self.hash_algorithm = hash_algorithm

    def hash_chain_generation(self, seed, chain_size):
        """
        Method to calculate an hash chain

        :param seed: Chain's seed (bytes)
        :param chain_size: Number of hashes
        :return: HashChain object
        """

        logging.debug(f'hash_chain_generation called with seed={seed} and chain_size={chain_size}')

        hash_chain = []

        to_digest = seed

        for i in range(1, chain_size + 1):
            md = hashlib.new(self.hash_algorithm)
            md.update(to_digest)

            to_digest = md.digest()
            to_digest = create_signed_byte_array_from(byte_array=to_digest)

            hash_chain.append(to_digest)

        return HashChain(hash_chain)
